pkg/cmd/server/kubernetes/node/options/options.go

```diff
@@ -147,10 +154,56 @@ func Build(options configapi.NodeConfig) (*kubeletoptions.KubeletServer, error)
 		server.HairpinMode = kubeletconfig.HairpinNone
 	}
 
+	server.RootDirectory, err = filepath.Abs(server.RootDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("unable to set absolute path for Kubelet root directory: %v", err)
+	}
+
 	return server, nil
 }
 
 func ToFlags(config *kubeletoptions.KubeletServer) []string {
 	server, _ := kubeletoptions.NewKubeletServer()
-	return cmdflags.AsArgs(config.AddFlags, server.AddFlags)
+	args := cmdflags.AsArgs(config.AddFlags, server.AddFlags)
+
+	// there is a special case.  If you set `--cgroups-per-qos=false` and `--enforce-node-allocatable` is
+	// an empty string, `--enforce-node-allocatable=""` needs to be explicitly set
+	if !config.CgroupsPerQOS && len(config.EnforceNodeAllocatable) == 0 {
+		args = append(args, `--enforce-node-allocatable=`)
```

Comment:
cgroups-per-qos isn't relevant to the special case; if `len(config.EnforceNodeAllocatable) == 0` then you should add `--enforce-node-allocatable=` regardless of the setting of `config.CgroupsPerQOS`. We just only noticed the problem in the `--cgroups-per-qos=false` case because the kubelet actually errors out in that case.

The issue is that the generic code isn't able to figure out the way that the default value of this argument works (ie, that the default value of `config.EnforceNodeAllocatable` when you don't say anything on the command line is `[]string{"pods"}`, not `[]string{}`). It seems like either this is a bug in the way that EnforceNodeAllocatable is defined, or else if not, then there are probably other options with similar default-value issues that ToFlags is mishandling too...



cgroups-per-qos only came up here because if CgroupsPerQOS is false, then you *have to* specify empty EnforceNodeAllocatable too, but in theory, someone might want to start a kubelet with CgroupsPerQOS true and empty EnforceNodeAllocatable.